### PR TITLE
Add thrift@0.10.0 w/ git auto-update

### LINKED ISSUE #8848: [Request] Add thrift
**Library name:** thrift
**Git repository url:**
**npm package url(optional):** 
**License(s):**
**Official homepage:**
**Wanna say something? Leave message here:**
# 

Notes from cdnjs maintainer:
You are welcome to add a library via sending pull request,
it'll be faster then just opening a request issue,
and please don't forget to read the guidelines for contributing, thanks!!

<bountysource-plugin>

---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/37333245-request-add-thrift?utm_campaign=plugin&utm_content=tracker%2F32893&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F32893&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by qwe19272375**
Pull request for issue: #8848 
Related issue(s): # #

Checklist for **Pull request** or **lib adding request issue** follows the conventions.

Note that if you are using a distribution purpose repository/package, please also provide the url and other related info like popularity of the source code repo/package.

# Profile of the lib
 * Git repository (required): https://git-wip-us.apache.org/repos/asf/thrift.git
 * Official website (optional, not the repository): https://thrift.apache.org/
 * NPM package url (optional): https://www.npmjs.com/package/thrift
 * License and its reference: Apache-2.0, [ref](https://github.com/apache/thrift/blob/master/LICENSE)
 * GitHub / Bitbucket popularity (required):
   - Count of watchers: 355
   - Count of stars: 3552
   - Count of forks: 2060
 * NPM download stats (optional):
   - Downloads in the last day: 1974
   - Downloads in the last week: 9889
   - Downloads in the last month: 45575

# Essential checklist
 * [ ] I'm the author of this library
   * [ ] I would like to add link to the page of this library on CDNJS on website / readme
 * [x] This lib was not found on cdnjs repo
 * [x] No already exist / duplicated issue and PR
 * [x] The lib has notable popularity
   * [x] More than 100 [Stars / Watchers / Forks] on [GitHub / Bitbucket]
   * [x] More than 500 downloads stats per month on npm registry
 * [x] Project has public repository on famous online hosting platform (or been hosted on npm)

# Auto-update checklist
 * [x] Has valid tags for each versions (for git auto-update)
 * [x] Auto-update setup
 * [x] Auto-update target/source is valid.
 * [x] Auto-update filemap is correct.

# Git commit checklist
 * [x] The first line of commit message is less then 50 chars, be clean and clear, easy to understand.
 * [x] The parent of the commit(s) in the PR is not old than 3 days.
 * [x] Pull request is sending from a non-master branch with meaningful name.
 * [x] Separate unrelated changes into different commits.
 * [x] Use rebase to squash/fixup dummy/unnecessary commits into only one commit.
 * [x] Close corresponding issue in commit message
 * [x] Mention related issue(s), people in commit message, comment.

close #8848, cc @jeking3

### CODE REVIEW

**Comment by qwe19272375** [comment]
The file structure have changed from `v0.9.2` and can't get the `thrift.js` file before it.
Should I change `fileMap` in `package.json` or add files manually?

**Comment by qwe19272375** [comment]
@cdnjs/intern3 Please help me review this PR, thanks.

File: ajax/libs/thrift/package.json [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+  ],
+  "repository": {
+    "type": "git",
+    "url": "https://git-wip-us.apache.org/repos/asf/thrift.git"
+  },
+  "version": "0.10.0",
+  "author": {
+    "name": "Apache Thrift Developers",
+    "email": "dev@thrift.apache.org",
+    "url": "http://thrift.apache.org"
```
Comment (sufuf3) [inline ajax/libs/thrift/package.json]: please use `https://thrift.apache.org/`
https is better when people browsing.

File: ajax/libs/thrift/0.10.0/thrift.min.js [hunk 2]
```diff
@@ -0,0 +1,2 @@
...hrift.MultiplexProtocol=function(t,e,i,r){Thrift.Protocol.call(this,e,i,r),this.serviceName=t},Thrift.inherits(Thrift.MultiplexProtocol,Thrift.Protocol,"multiplexProtocol"),Thrift.MultiplexProtocol.prototype.writeMessageBegin=function(t,e,i){e===Thrift.MessageType.CALL||e===Thrift.MessageType.ONEWAY?Thrift.Protocol.prototype.writeMessageBegin.call(this,this.serviceName+":"+t,e,i):Thrift.Protocol.prototype.writeMessageBegin.call(this,t,e,i)},Thrift.Multiplexer=function(){this.seqid=0},Thrift.Multiplexer.prototype.createClient=function(t,e,i){e.Client&&(e=e.Client);var r=this;e.prototype.new_seqid=function(){return r.seqid+=1,r.seqid};var s=new e(new Thrift.MultiplexProtocol(t,i));return s};var copyList,copyMap;copyList=function(t,e){if(!t)return t;var i;i=void 0===e.shift?e:e[0];var r,s,n=i,o=t.length,h=[];for(r=0;r<o;r++)s=t[r],null===i?h.push(s):i===copyMap||i===copyList?h.push(i(s,e.slice(1))):h.push(new n(s));return h},copyMap=function(t,e){if(!t)return t;var i;i=void 0===e.shift?e:e[0];var r,s=i,n={};for(var o in t)t.hasOwnProperty(o)&&(r=t[o],null===i?n[o]=r:i===copyMap||i===copyList?n[o]=i(r,e.slice(1)):n[o]=new s(r));return n},Thrift.copyMap=copyMap,Thrift.copyList=copyList;
```
Comment (jeking3) [inline ajax/libs/thrift/0.10.0/thrift.min.js]: Should this say 0.9.3 or 0.10.0?

**Comment by sufuf3** [comment]
@qwe19272375 could you discuss with the author about the `thrift.min.js` of v0.10.0?
And your commit message about cc the author, I think you could refer to https://www.npmjs.com/package/thrift .
Thanks.

**Comment by qwe19272375** [comment]
It seems that I misplaced `0.9.3/thrift.js` as `0.10.0/thrift.js` and I fixed it.
About cc the author, I tagged the person in [apache/thrift](https://github.com/apache/thrift) git repository who has the most commits in the past year.

File: ajax/libs/thrift/package.json [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+  "version": "0.10.0",
+  "author": {
+    "name": "Apache Thrift Developers",
+    "email": "dev@thrift.apache.org",
+    "url": "https://thrift.apache.org"
+  },
+  "license": "Apache-2.0",
+  "autoupdate": {
+    "source": "git",
+    "target": "https://git-wip-us.apache.org/repos/asf/thrift.git",
```
Comment (clairetsai818) [inline ajax/libs/thrift/package.json]: You should use `git://` instead of `https://`

Comment (sufuf3) [inline ajax/libs/thrift/package.json]: @clairetsai818 I use `git clone git://git-wip-us.apache.org/repos/asf/thrift.git` but I can't download the library. Did you use it successfully?
What I get the message as the followings:
```
Cloning into 'thrift'...
fatal: unable to connect to git-wip-us.apache.org:
git-wip-us.apache.org[0: 140.211.11.23]: errno=Connection refused
```

Comment (extend1994) [inline ajax/libs/thrift/package.json]: I think for auto-update mechanism both are acceptable. I run auto-update and both succeed.

File: ajax/libs/thrift/package.json [hunk 4]
```diff
@@ -0,0 +1,36 @@
+{
+  "name": "thrift",
+  "filename": "thrift.min.js",
+  "description": "The Apache Thrift software framework, for scalable cross-language services development, combines a software stack with a code generation engine to build services that work efficiently and seamlessly between C++, Java, Python, PHP, Ruby, Erlang, Perl, Haskell, C#, Cocoa, JavaScript, Node.js, Smalltalk, OCaml and Delphi and other languages.",
```
Comment (extend1994) [inline ajax/libs/thrift/package.json]: This sentence is from homepage, which is different from `package.json/description` in the upstream. Not sure if we can or should change it.

**Comment by qwe19272375** [comment]
Split to two commits and modified `description` in `package.json`,
@cdnjs/intern3 please help me review it, thanks!

**Comment by sufuf3** [comment]
```
Add v0.3.0~v0.9.1 from `lib/js/thrift.js` because git auto-update can't 
get them.
```
please change to 
```
Add v0.3.0~v0.9.1 from GitHub due to different file structure.
```

`lib/js/thrift.js` should fill in where did you get the files (GitHub or npm).
And we know that you add old versions manually because git auto-update can't get them.
If you descript more clear about why git auto-update can't get them, such as file structure, the file named... etc, this will help us know more clear.

**Comment by sufuf3** [comment]
I saw that `thrift.js` both in `lib/nodejs/lib/thrift/thrift.js` and `lib/js/src/thrift.js`.
How did you know that CDNJS should host `thrift.js` which under `lib/js/src/` fodler?

**Comment by qwe19272375** [comment]
Opened issue to discuss which files should be added.
https://issues.apache.org/jira/browse/THRIFT-4194

File: ajax/libs/thrift/package.json [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+  },
+  "license": "Apache-2.0",
+  "autoupdate": {
+    "source": "git",
+    "target": "https://git-wip-us.apache.org/repos/asf/thrift.git",
+    "fileMap": [
+      {
+        "basePath": "lib/js/src",
+        "files": [
+          "thrift*"
```
Comment (sufuf3) [inline ajax/libs/thrift/package.json]: @qwe19272375 sorry about that  I was not familiar with nodejs in the past.
I think that using this path is fine.
But the `bathPath` should be `lib`
And the `files` should be `"js/src/thrift*"`

Then every thing looks good to me.
Thank you.

**Comment by sufuf3** [comment]
@qwe19272375 would you fix the PR after the final exam?
this will be a great work.
Many thanks.

**Comment by sashberd** [comment]
@sufuf3 @extend1994 @PeterDaveHello Cannot rebase this PR to latest one. Suggest to close.

**Comment by extend1994** [comment]
@sashberd Take you time

**Comment by sashberd** [comment]
Moved to #13694

**Pull request closed without merging**